### ISSUE TITLE
HTTP/2 and Protobuf are not gRPC exclusive

### ISSUE THREAD

**Issue by safakgur** [dotnet/AspNetCore.Docs#18083]
The documentation here implies that HTTP APIs don't have support for HTTP/2 and that they all use JSON as their message exchange format while there are a lot of HTTP APIs that use HTTP/2 with Protobuf out there. It also implies that gRPC always uses Protobuf while it actually supports any format (e.g. JSON) just like a regular HTTP API.

So even though the page claims to compare gRPC services to HTTP APIs, it mostly compares HTTP to HTTP/2 and JSON to Protobuf. I suggest comparing gRPC services using Protobuf to HTTP APIs using HTTP/2 and Protobuf to highlight the actual differences.

Stating that Protobuf is the default IDL for gRPC is fine. Stating that most frameworks use JSON as the default format for HTTP APIs is also fine, as long as the document doesn't imply that gRPC=HTTP/2+Protobuf and HTTP=HTTP/1.1+JSON.

For reference, gRPC FAQ has a ["Why is gRPC better than any binary blob over HTTP/2?"](https://grpc.io/faq/#why-is-grpc-better-than-any-binary-blob-over-http2) question with the answer:

> **This is largely what gRPC is on the wire.** However gRPC is also a set of libraries that will provide higher-level features consistently across platforms that common HTTP libraries typically do not.
> Examples of such features include:
>
> * interaction with flow-control at the application layer
> * cascading call-cancellation
> * load balancing & failover

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8321e4f0-1d12-c4cd-76af-a7df9ca2e732
* Version Independent ID: c9e71059-2e31-3de3-eb26-0946cfe0ec3d
* Content: [Compare gRPC services with HTTP APIs](https://docs.microsoft.com/en-us/aspnet/core/grpc/comparison?view=aspnetcore-3.1#feedback)
* Content Source: [aspnetcore/grpc/comparison.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/grpc/comparison.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-grpc**
* GitHub Login: @JamesNK
* Microsoft Alias: **jamesnk**

**Comment by wadepickett** [comment]
Thanks @safakgur for taking the time to detail your suggestions for improving this topic.  Greatly appreciated.  

@JamesNK, thoughts on this one before I edit?

**Comment by JamesNK** [comment]
The point of the documentation is to compare what developers are familiar with, JSON APIs, with what is new: gRPC. Comparing gRPC to something that is uncommon isn't useful.

The page says it is comparing gRPC to HTTP APIs with JSON in a number of places.

It can be updated to mention somewhere that HTTP/2 and its benefits can be used together with HTTP APIs with JSON.

**Comment by safakgur** [comment]
**tldr;** Liked your additions on HTTP/2, wish you added some on Protobuf too.

---

Thanks for the additions, it looks much clear now that HTTP/2 is available for HTTP APIs as well.

> The point of the documentation is to compare what developers are familiar with, JSON APIs, with what is new: gRPC.

I agree with where you're coming from (comparing the familiar with the new) but like you said, what is new is gRPC. Protobuf came out more than a decade ago and we could always use it on HTTP APIs. So seeing Protobuf under the "gRPC Strengths" without any disclaimers like the ones you added about HTTP/2 is, at least in my opinion, a bit misleading.

**For context:** I used HTTP/2 and Protobuf for HTTP APIs before (I never thought it being so uncommon, but I'll trust you on that) but I had never used gRPC. So to get a quick comparison, I did a [Google search][1] that looked like this:

![image](https://user-images.githubusercontent.com/1020499/81279389-fb870480-904e-11ea-9811-c508153f3b6c.png)

...which led me to this page that, before your update, looked more about "Protobuf vs JSON" and "HTTP/2 vs HTTP" than "gRPC services vs HTTP APIs".

Sorry for the longish comment, I just wanted to be clear about my thought process.

[1]: https://www.google.com/search?q=grpc+vs+http
